mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp

```diff
@@ -1907,6 +1907,20 @@ void BroadcastOp::getEffects(
                         getDpsInits());
 }
 
+LogicalResult BroadcastOp::canonicalize(BroadcastOp op,
```

Comment:
We can keep this, but a similar optimization is written for `linalg.generic` that is more general. This is fine for me, but maybe moving the https://github.com/llvm/llvm-project/blob/659419ac2f05035cd8ff19806a9e48ccc725d3ba/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L1095 to work on the `LinalgOp` interface will just make all of this unnecessary. I am not strongly opinionated though. This looks fine as well.

Reply:
yah i was actually thinking that doing this in the interface would be better. that way we don't have to keep doing this for other named ops.

Reply:
hmm, i guess not since `OpInterface` doesn't seem to have `hasCanonicalizer` or `hasCanonicalizationMethod`.  I'm sure there are ways to do it.  I just dont know how. But i'd rather figure that out, if reasonable, than have to do this for other named ops later.

Reply:
I have a dirty trick. The `getCanonicalization` is just a hook. It could bring in any pattern. You cuold just make that pattern work on `LinalgOp` and it will get pulled into canonicalizer and apply to any LinalgOp. (I said it was a dirty trick)

Reply:
I dont mind dirty tricks as long as they are acceptable and not vulnerable to issues. :)

Reply:
In this case, not so dirty.  A `linalg.generic` is just a "generic" version of all `LinalgOp`s. So any pattern on `linalg.generic` is as good a pattern on `LinalgOp`s. So from my perspective, in this particular case, its actually not bad. 
Ideally we should have a way to have canonicalizers on interfaces as well. Would be neat!

Reply:
i'm not exactly sure how to get the canonicalizer pass to grap `LinalgOp::getCanonicalizationPattern`.  I naively added it to `extraClassDeclarations` in `LinalgStructuredInterface` and then just replaced `GenericOp::getCanonicalizationPattern`. Any ideas?

Reply:
The only thing i can think of is just to have this, but still require ops to add their own canonicalizer at which point can reuse this same pattern

Reply:
I took a crack at it. First I tried to do an `OpInterfaceRewriterPatter<LinalgOp>` but that ended up overriding all other linalg op's canonicalizers. So i opted for a templated rewrite

Reply:
any thoughts? :)